### ISSUE TITLE
Conan prioritizes indirect dependency over direct dependency

### ISSUE THREAD

**Issue by monsdar** [conan-io/conan#5900]
I'm having the following component tree:

* LibConsumer uses LibCore/1.0.0+abc@user/channel
* MyComp uses LibConsumer and LibCore/1.0.0+xyz@user/channel

conanfile.py in MyComp:

```
...
    def requirements(self):
        self.requires( "LibConsumer/1.0.0@user/channel") #indirectly uses LibCore/1.0.0+abc@user/channel
        self.requires( "LibCore/1.0.0+xyz@user/channel")
...
```

When running a `conan install` on MyComp Conan resolves LibCore to `LibCore/1.0.0+abc@user/channel`. This is probably due to LibConsumer being first in the list of requires and the SemVer of both LibCore packages not being different. When I swap the requires-statement to the following Conan resolves LibCore to `LibCore/1.0.0+xyz@user/channel`:

```
...
    def requirements(self):
        self.requires( "LibCore/1.0.0+xyz@user/channel")
        self.requires( "LibConsumer/1.0.0@user/channel")
...
```

I would have expected Conan to always prefer the direct dependencies over the indirect dependencies, thus always resolving to `LibCore/1.0.0+xyz@user/channel` because it has explicitly been named. I know that there isn't a difference between from a SemVer viewpoint, but there could be reasons why someone put the one over the other into the conanfile.

Is there a way to enforce using the direct deps over the indirect ones? Should this be changed by default?

**Comment by lasote** [comment]
I've been able to reproduce. The reason is that Conan compares the versions and by definition, the `Build metadata` (+xyz) is discarded, so it considers both references equal. Nevertheless, if the reference is not the same (and the package either) I think this behavior is not correct.
I would try to propose a fix but I'm a bit afraid it can break something else.

**Comment by monsdar** [comment]
Right now it breaks our builds at least :/

Sometimes our CI rebuilds a package and changes a dependency (we're using aliases). The version of the package hasn't changed, but it is "something else" than before. Therefore we add the Git SHA to the metadata. Sometimes a project wants to use a specific version of a package, down to the metadata level.
We could fix this by increasing the version number everytime we make a change. But that'd be quite complex. The CI would need to increase the version number automatically, recommit this into the repo etc.

Would it be possible to fix this with a config value? Something like `preferExplicitRequirements = True` or `doNotIgnoreSemverMetadata = True`?

**Comment by lasote** [comment]
I'm not saying it is not going to be fixed, I'm investigating right now, I'll update asap.

**Comment by monsdar** [comment]
Got ya! Just wanted to give a bit more context.

**Comment by memsharded** [comment]
Fixed in #5903 will be released in 1.19.2 soon. Thanks for reporting!
